waku/v2/utils/peer.go

```diff
@@ -37,5 +44,75 @@ func SelectPeer(host host.Host, protocolId string) (*peer.ID, error) {
 		return &peers[rand.Intn(len(peers))], nil // nolint: gosec
 	}
 
-	return nil, errors.New("no suitable peers found")
+	return nil, ErrNoPeersAvailable
+}
+
+type pingResult struct {
+	p   peer.ID
+	rtt time.Duration
+}
+
+func SelectPeerWithLowestRTT(ctx context.Context, pingService *ping.PingService, protocolId string) (*peer.ID, error) {
+	if pingService == nil {
+		return nil, PingServiceNotAvailable
+	}
+
+	var peers peer.IDSlice
+	for _, peer := range pingService.Host.Peerstore().Peers() {
+		protocols, err := pingService.Host.Peerstore().SupportsProtocols(peer, protocolId)
+		if err != nil {
+			log.Error("error obtaining the protocols supported by peers", err)
+			return nil, err
+		}
+
+		if len(protocols) > 0 {
+			peers = append(peers, peer)
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	waitCh := make(chan struct{})
+	pingCh := make(chan pingResult, 1000)
+
+	go func() {
+		for _, p := range peers {
+			wg.Add(1)
+			go func(p peer.ID) {
+				defer wg.Done()
+				ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+				defer cancel()
+				result := <-pingService.Ping(ctx, p)
```

Comment:
Coud we not build the ping service before the go func (line 79) that would avoid to pass it down, I also imagine that this is quite simple lib? Maybe you want to offer to choose which ping service to use?

Reply:
it seems that i can use Ping function directly without having to instantiate the pingService.  https://github.com/libp2p/go-libp2p/blob/master/p2p/protocol/ping/ping.go#L84-L130  
I'll update the code to avoid passing a ping service!